src/bonjour.c

```diff
@@ -0,0 +1,102 @@
+#include <dns_sd.h>
+#include <stdio.h>
+#include "bonjour.h"
+#include "lxi.h"
+
+char *services[] = {
+    "_lxi._tcp",
+    "_vxi-11._tcp",
+};
+
+void resolve_callback(
+    DNSServiceRef sdRef,
+    DNSServiceFlags flags,
+    uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode,
+    const char *fullname,
+    const char *hosttarget,
+    uint16_t port,
+    uint16_t txtLen,
+    const unsigned char *txtRecord,
+    void *context
+) {
+    printf("Resolve callback\n");
+    printf("Full name: %s\n", fullname);
+    printf("Host target: %s\n", hosttarget);
+    printf("Port: %u\n\n", ntohs(port));
+
+    lxi_info_t *info = (lxi_info_t *)context;
+    info->service(fullname, hosttarget, "Resolved service", ntohs(port));
+}
+
+void browse_callback(
+    DNSServiceRef service,
+    DNSServiceFlags flags,
+    uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode,
+    const char *serviceName,
+    const char *regtype,
+    const char *replyDomain,
+    void *context
+) {
+    printf("Browse callback\n");
+    printf("Service name: %s\n", serviceName);
+    printf("Regtype: %s\n", regtype);
+    printf("Domain: %s\n\n", replyDomain);
+
+    lxi_info_t *info = (lxi_info_t *)context;
+    info->broadcast(serviceName, "Found service");
+
+    DNSServiceRef resolveService;
+    DNSServiceErrorType error;
+
+    error = DNSServiceResolve(&resolveService, 0, interfaceIndex, serviceName, regtype, replyDomain, resolve_callback, info);
+    if (error != kDNSServiceErr_NoError) {
+        fprintf(stderr, "DNSServiceResolve() failed: %d\n", error);
+        return;
+    }
+
+    error = DNSServiceProcessResult(resolveService);
+    if (error != kDNSServiceErr_NoError) {
+        fprintf(stderr, "DNSServiceProcessResult() failed: %d\n", error);
+        return;
+    }
+
+    DNSServiceRefDeallocate(resolveService);
+}
+
+int bonjour_discover(lxi_info_t *info, int timeout) {
+    DNSServiceRef service;
+    DNSServiceErrorType error;
+    int i;
+
+    for(i = 0; i < 5; i++) {
+        error = DNSServiceBrowse(&service, 0, 0, services[i], "local.", browse_callback, info);
+        if(error != kDNSServiceErr_NoError) {
+            fprintf(stderr, "DNSServiceBrowse() failed: %d\n", error);
+            return LXI_ERROR;
+        }
+
+        error = DNSServiceProcessResult(service);
+        if(error != kDNSServiceErr_NoError) {
+            fprintf(stderr, "DNSServiceProcessResult() failed: %d\n", error);
+            return LXI_ERROR;
+        }
+
+        DNSServiceRefDeallocate(service);
+        return LXI_OK;
```

Comment:
This line is the reason your code is only searching for the first service in your services array. You don't want to return from the function here. Simply move that down to the end of your function and you should be fine.